prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+   the peer does not support it (this can happen when endpoint is using libfabric 1.12,
+   but the peer is using libfabric 1.10), the requester need to return an error to the
+   application (section 4.2)
+
+Regarding extra request, if an endpoint can support an extra request the peer has, it
+should comply the request. Otherwise, it can ignore the request. Peer should be do
+one of the following:
+
+- a. carry on the communication without using the extra request. (see section 4.4
+     for example)
+
+- b. abort the communication and return an error to application. (see section 4.3
+     for example)
+
+For example, if sender is using libfabric 1.10, and receiver is using 1.14. Receiver
+will have the extra request "connid header", but sender does not support it. In this
+case, sender should not do anything special, it is receiver's responsibility to react
+accordingly. (section 4.4)
+
+This concludes the workflow of the handshake sub-protocol.
+
+The binary format of a HANDSHAKE packet is listed in table 2.2.
+
+Table: 2.2 binary format of the HANDSHAKE packet
+
+| Name      | Length (bytes) | type | C language type |
+|-|-|-|-|
+| `type`    | 1 | integer | `uint8_t`  |
+| `version` | 1 | integer | `uint8_t`  |
+| `flags`   | 2 | integer | `uint16_t` |
+| `nex_p3`  | 4 | integer | `uint32_t` |
+| `exinfo`  | `8 * (nex_p3 - 3)` | integer array | `uint64_t[]` |
+
+The first 4 bytes (3 fields: `type`, `version`, `flags`) is the EFA RDM base header (section 1.3).
+
+Immediately after the base header, there are 2 fields `nex_p3` and `exinfo`. 
+
+The field `exinfo` is an array of 8 byte integers, which stores the capability of an endpoint.
+
+As mentioned before, each extra feature/request was assigned an ID when it was introduced to protocol v4.
+When constructing the handshake packet, for each extra feature it supports (or an extra request it want to impose),
+an endpoint need to toggle on a corresponding bit in the `exinfo` array. Specifically, if an endpoint supports
+the extra feature with ID `i` (or want to impose extra request with ID `i`), it needs to toggle on the
+No. `i%64` bit of the No. `i/64` member of the `exinfo` array.
+
+For example, if an endpoint supports the extra feature "RDMA read based data transfer" (ID 0), it needs to
+toggle on the No. 0 (the first) bit of `exinfo[0]`. (section 4.1)
+
+If an endpoint wants to impose the "stable header length" extra request it need to toggle on bit No 2.
+in `exinfo[0]`. (section 4.3)
+
+Note, the field `exinfo` was named `features` when protocol v4 was initially introduced, at that time we
+only planned for extra features. Later, we discovered that the handshake sub-protocol can also be used to pass
+additional request information, thus introduced the concept of "extra request" and renamed this field `exinfo`.
+
+`nex_p3` is number of `exinfo` flags of the endpoint plus 3. The "plus 3" is for historical reasons.
```

Comment:
Historical reasons? Or backwards compatibility? I'm a bit lost here with the changes to this packet.